README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->
Flutter Horizontal Date  Picker Library that provides a calendar  center selection of items.

<p>
 <img src="https://github.com/SaNu-hIT/Flutter-Horizontal-Calendar/blob/main/Calendar.jpg"/>
</p>


## How To Use


Use this package as a library

```
dependencies:
     horizontalcalender: ^0.0.4
```


Import the following package in your dart file

```dart
import 'package:horizontalcalendar/horizontalcalendar.dart';
```

## Usage

Use the `HorizontalCalendar` Widget

```dart
  @override
  Widget build(BuildContext context) {
    final FixedExtentScrollController itemController =
        FixedExtentScrollController();
    return Container(
        height: 200,
        width: MediaQuery.of(context).size.width,
        padding: EdgeInsets.all(10.0),
        child: HorizontalCalendar(DateTime.now(),
            width: MediaQuery.of(context).size.width*.25,
            height: 120,
            selectionColor: Colors.red,
            itemController: itemController));
  }
```

##### Constructor:

```dart
  HorizontalCalendar(
    this.startDate, {
    Key? key,
    this.width = 40,
    this.height = 80,
    this.controller,
    this.monthTextStyle,
    this.dayTextStyle,
    this.dateTextStyle,
    this.selectedTextColor,
    this.selectionColor,
    this.deactivatedColor,
    this.initialSelectedDate,
    this.activeDates,
    this.inactiveDates,
    this.daysCount = 500,
    this.onDateChange,
    this.locale = "en_US",
    this.selectedDayStyle,
    this.selectedDateStyle,
    required this.itemController,
  });
```


CHANGELOG.md
## calendar - Release

* Beta version release

## 0.0.2 - Release

* Bug fixes


## 0.0.3 - Release

* More examples and documentation

## 0.0.4 - Release

* Update Package Structure
*



lib/date_item_widget.dart
import 'package:auto_size_text/auto_size_text.dart';
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import 'listener/tap.dart';

class DateItemWidget extends StatelessWidget {
  final double? width;
  final DateTime date;
  final TextStyle? monthStyle, dayStyle, dateStyle;
  final Color selectionColor;
  final DateSelectionCallback? onDateSelected;
  final String? locale;

  const DateItemWidget({
    required this.date,
    required this.monthStyle,
    required this.dayStyle,
    required this.dateStyle,
    required this.selectionColor,
    this.width,
    this.onDateSelected,
    this.locale,
  });

  @override
  Widget build(BuildContext context) {
    return InkWell(
      child: Container(
        width: width,
        margin: EdgeInsets.all(3.0),
        decoration: BoxDecoration(
          borderRadius: BorderRadius.all(Radius.circular(8.0)),
          color: selectionColor,
        ),
        child: Padding(
          padding: EdgeInsets.all(8),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: <Widget>[
              AutoSizeText(
                  new DateFormat("MMM", locale).format(date).toUpperCase(),
                  // Month
                  style: monthStyle),
              AutoSizeText(date.day.toString(), // Date
                  style: dateStyle),
              AutoSizeText(
                  new DateFormat("E", locale).format(date).toUpperCase(),
                  // WeekDay
                  style: dayStyle)
            ],
          ),
        ),
      ),
      onTap: () {
        // Check if onDateSelected is not null
        if (onDateSelected != null) {
          // Call the onDateSelected Function
          onDateSelected!(this.date);
        }
      },
    );
  }
}


lib/horizontalcalendar.dart
library horizontalcalendar;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:horizontalcalender/date_item_widget.dart';

import 'extra/color.dart';
import 'extra/style.dart';
import 'listener/tap.dart';

class HorizontalCalendar extends StatefulWidget {
  final DateTime startDate;

  final double width;

  final double height;

  final DatePickerController? controller;

  final Color selectedTextColor;

  final Color selectionColor;

  final Color deactivatedColor;

  final TextStyle monthTextStyle;

  final TextStyle dayTextStyle;

  final TextStyle selectedDayStyle;

  final TextStyle selectedDateStyle;

  final TextStyle dateTextStyle;

  final DateTime?  initialSelectedDate;

  final List<DateTime>? inactiveDates;

  final List<DateTime>? activeDates;

  final DateChangeListener? onDateChange;

  final int daysCount;

  final String locale;

  final FixedExtentScrollController itemController;

  HorizontalCalendar(
    this.startDate, {
    Key? key,
    this.width = 40,
    this.height = 80,
    this.controller,
    this.monthTextStyle = defaultMonthTextStyle,
    this.dayTextStyle = defaultDayTextStyle,
    this.dateTextStyle = defaultDateTextStyle,
    this.selectedTextColor = Colors.white,
    this.selectionColor = AppColors.defaultSelectionColor,
    this.deactivatedColor = AppColors.defaultDeactivatedColor,
    this.initialSelectedDate,
    this.activeDates,
    this.inactiveDates,
    this.daysCount = 500,
    this.onDateChange,
    this.locale = "en_US",
    this.selectedDayStyle = defaultSelectedTextStyle,
    this.selectedDateStyle = defaultSelectedTextStyle,
    required this.itemController,
  });

  @override
  State<StatefulWidget> createState() => _HorizontalCalendarState();
}

class _HorizontalCalendarState extends State<HorizontalCalendar> {
  ScrollController _controller = ScrollController();
  DateTime? _currentDate;

  late final TextStyle selectedDateStyle;
  late final TextStyle selectedMonthStyle;
  late final TextStyle selectedDayStyle;
  late final TextStyle deactivatedDateStyle;
  late final TextStyle deactivatedMonthStyle;
  late final TextStyle deactivatedDayStyle;
  var itemSelected = 0;

  @override
  Widget build(BuildContext context) {
    return Container(
      height: widget.height,
      child: Column(
        children: [
          Container(
            height: widget.height,
            child: RotatedBox(
                quarterTurns: -1,
                child: ListWheelScrollView(
                  diameterRatio: 10,
                  // useMagnifier: true,
                  offAxisFraction: 0,
                  onSelectedItemChanged: (x) {
                    setState(() {
                      itemSelected = x;
                      print("selected");
                      print(x);

                      DateTime _date = widget.startDate.add(Duration(days: x));
                      DateTime date =
                          new DateTime(_date.year, _date.month, _date.day);
                      if (widget.onDateChange != null) {
                        widget.onDateChange!(date, x);
                      }
                    });
                  },
                  children: List.generate(
                      widget.daysCount,
                      (x) => RotatedBox(
                          quarterTurns: 1,
                          child: AnimatedContainer(
                              duration: Duration(milliseconds: 100),
                              child: StreamBuilder<Object>(
                                  stream: null,
                                  builder: (context, snapshot) {
                                    DateTime date;
                                    DateTime _date =
                                        widget.startDate.add(Duration(days: x));
                                    date = new DateTime(
                                        _date.year, _date.month, _date.day);
                                    bool isDeactivated = false;
                                    // check if this date needs to be deactivated for only DeactivatedDates
                                    if (widget.inactiveDates != null) {
                                      for (DateTime inactiveDate
                                          in widget.inactiveDates!) {
                                        if (_compareDate(date, inactiveDate)) {
                                          isDeactivated = true;
                                          break;
                                        }
                                      }
                                    }
                                    // check if this date needs to be deactivated for only ActivatedDates
                                    if (widget.activeDates != null) {
                                      isDeactivated = true;
                                      for (DateTime activateDate
                                          in widget.activeDates!) {
                                        // Compare the date if it is in the
                                        if (_compareDate(date, activateDate)) {
                                          isDeactivated = false;
                                          break;
                                        }
                                      }
                                    }

                                    // Check if this date is the one that is currently selected
                                    // bool isSelected =
                                    // _currentDate != null ? _compareDate(date, _currentDate!) : false;
                                    // Check if this date is the one that is currently selected
                                    bool isSelected = itemSelected == x;
                                    // Return the Date Widget
                                    return DateItemWidget(
                                      date: date,
                                      monthStyle: isDeactivated
                                          ? deactivatedDayStyle
                                          : isSelected
                                              ? widget.selectedDayStyle
                                              : widget.dayTextStyle,
                                      dateStyle: isDeactivated
                                          ? deactivatedDateStyle
                                          : isSelected
                                              ? widget.selectedDateStyle
                                              : widget.dateTextStyle,
                                      dayStyle: isDeactivated
                                          ? deactivatedDayStyle
                                          : isSelected
                                              ? widget.selectedDayStyle
                                              : widget.dayTextStyle,
                                      width: widget.width,
                                      locale: widget.locale,
                                      selectionColor: isSelected
                                          ? widget.selectionColor
                                          : Colors.white,
                                      onDateSelected: (selectedDate) {
                                        // Don't notify listener if date is deactivated
                                        if (isDeactivated) return;

                                        // A date is selected
                                        // if (widget.onDateChange != null) {
                                        //   widget.onDateChange!(selectedDate);
                                        // }
                                        // if (isSelected) {
                                        //   widget.onDateChange!(selectedDate);
                                        // }

                                        setState(() {
                                          _currentDate = selectedDate;
                                        });
                                      },
                                    );
                                  })))),
                  itemExtent: widget.width,
                )),
          ),
        ],
      ),
    );
  }

  bool _compareDate(DateTime date1, DateTime date2) {
    return date1.day == date2.day &&
        date1.month == date2.month &&
        date1.year == date2.year;
  }
}

class DatePickerController {
  _HorizontalCalendarState? _datePickerState;

  void setDatePickerState(_HorizontalCalendarState state) {
    _datePickerState = state;
  }

  void jumpToSelection() {
    assert(_datePickerState != null,
        'DatePickerController is not attached to any DatePicker View.');

    // jump to the current Date
    _datePickerState!._controller
        .jumpTo(_calculateDateOffset(_datePickerState!._currentDate!));
  }

  /// This function will animate the Timeline to the currently selected Date
  void animateToSelection(
      {duration = const Duration(milliseconds: 500), curve = Curves.linear}) {
    assert(_datePickerState != null,
        'DatePickerController is not attached to any DatePicker View.');

    // animate to the current date
    _datePickerState!._controller.animateTo(
        _calculateDateOffset(_datePickerState!._currentDate!),
        duration: duration,
        curve: curve);
  }

  /// This function will animate to any date that is passed as a parameter
  /// In case a date is out of range nothing will happen
  void animateToDate(DateTime date,
      {duration = const Duration(milliseconds: 500), curve = Curves.linear}) {
    assert(_datePickerState != null,
        'DatePickerController is not attached to any DatePicker View.');

    _datePickerState!._controller.animateTo(_calculateDateOffset(date),
        duration: duration, curve: curve);
  }

  /// Calculate the number of pixels that needs to be scrolled to go to the
  /// date provided in the argument
  double _calculateDateOffset(DateTime date) {
    final startDate = new DateTime(
        _datePickerState!.widget.startDate.year,
        _datePickerState!.widget.startDate.month,
        _datePickerState!.widget.startDate.day);

    int offset = date.difference(startDate).inDays;
    return (offset * _datePickerState!.widget.width) + (offset * 6);
  }
}
